### DISCUSSION TITLE
How to use tsTypes in project with ES Modules and nodenext moduleResolution?

### DISCUSSION THREAD

**Discussion by tomg7** [statelyai/xstate#3933] [Q&A] [answered]
I am trying to upgrade my project to ES Modules.
I have set `"type": "module"` in package.json and `"moduleResolution": "nodenext"` in tsconfig.

```
const machine = createMachine({
   tsTypes: {} as import("./index.typegen").Typegen0,
```

in `index.ts` gives "Cannot find module './index.typegen' or its corresponding type declarations." error as it expects the .js extension.

I tried to add the .js extension but the VS Code Extension removes it when I save the file.

Is there a way for me to add tsTypes with the above settings or is it something that needs to be handled in XState?

**Comment by Andarist** [comment, marked as answer]
This is currently not quite possible. We need to make typegen part of things more configurable. The plan is to add support for `tsTypes: {} as import("./index.typegen.d.ts").Typegen0,` (note the `.d.ts` part here) to the TypeScript itself (I was told by the TS team that this should be supported but it's currently not). With that in place we could just add an explicit `.d.ts` extension regardless of your settings and it should just work. We need to evaluate our options for pre-5.1 users though.

**Reply by tomg7** [reply]
Hi Andarist, thanks for the detailed explanation. Looks like the ESM update of my project has to wait a little :)

**Reply by andreash** [reply]
I am having the same problem. How about simply putting an option into the VS Code extension to extend the import filename with ".js"? Sound to me like this may be simple to do and enables the use of ES Modules - which has other advantages.

**Reply by Andarist** [reply]
Yea, we could allow an explicit config like that today as an opt-in. In the future the default would be to use an explicit `.d.ts` extension automatically. We'd gladly accept a PR introducing such an option.

**Reply by andreash** [reply]
:) - huh, yaix. I admit I haven't got my feet wet yet on coding VS Code extensions. Anybody volunteering?

**Reply by andreash** [reply]
I have found the code - are there any description how to compile and debug the extension? Then I will try to setup a PR.

**Reply by andreash** [reply]
Sunday morning hack - found out how it works and will send you the PR.

**Reply by JosephDenman** [reply]
I'm having this issue right now. Could you share the hack?

**Reply by Andarist** [reply]
You might try to use `useDeclarationFileForTypegenData` setting in our VS Code extension
